### ISSUE TITLE
[feature request:] expose `cum_log_probs` as one of results for gpt/decoder?

### ISSUE THREAD

**Issue by gyin-ai** [NVIDIA/FasterTransformer#84]
Can I ask whether we can return and expose the `cum_log_probs` as one of results? Thanks. For example, lightseq and ctranslate2 would expose them as one of results which is quite useful for downstream tasks during the ranking or threshold. 

1. lightseq: https://github.com/bytedance/lightseq/blob/b2b99bb2559b804c8b2e277465227873f5696baf/server/decoder_generate_server.cc.cu#L441
2. ctranslate2: 
https://github.com/OpenNMT/CTranslate2/blob/85a8a41aae84df2979b06fd004e67f92ef9d4760/python/translator.cc#L224

**Comment by byshiue** [comment]
Currently, FT does not support. 
we have provide the ranking function like beam search and sampling. Do you need the cum_log_probs to other tasks? 
Besides, FT decoding generates all tokens at once, we should return the cum_log_probs of which step?

**Comment by gyin-ai** [comment]
> Currently, FT does not support.
> we have provide the ranking function like beam search and sampling. Do you need the cum_log_probs to other tasks?
> Besides, FT decoding generates all tokens at once, we should return the cum_log_probs of which step?

Sorry. Just want to clarify, the cum_log_probs is the final output for all generated tokens. And ranking or threshold means other downstream tasks that take the generated sequence and score as the input, not within each auto-regressive step.

**Comment by byshiue** [comment]
We will consider this suggestion.
Since FasterTransformer focuses on the efficiency of general case, adding these features may hurt its performance or requires more memory. For example, if the sequence length is 1024, batch size is 32, and the vocab size is 51200, the log_prob of all steps requires 6 GBs under FP32.

**Comment by gyin-ai** [comment]
> We will consider this suggestion.
> Since FasterTransformer focuses on the efficiency of general case, adding these features may hurt its performance or requires more memory. For example, if the sequence length is 1024, batch size is 32, and the vocab size is 51200, the log_prob of all steps requires 6 GBs under FP32.

The cum_log_prob score is just the score for predicted tokens, not the score of logits for all vocab tokens. Hence the memory usage would be one float number if we cumulative all of them or at most predicted tokens*one float number.

**Comment by byshiue** [comment]
Because you mention `log_prob` in """Another option would be to expose a list of `log_probs` for each token when the full sequence generation is finished. The downstream user can choose how they use the score, either cumulative, mean or more sophisticated operations."""
Only return the `cum_log_prob` is possible, but need to check because we do some optimization and we may not have the full results of `cum_log_prob`. Besides, is this request only focusing on beam search?

**Comment by gyin-ai** [comment]
Thanks. Can we also add it as the sampling decoding output? I saw gpt example used sampling. 

I noticed we had something mentioned here. https://github.com/NVIDIA/FasterTransformer/blob/93b7751bcbb4b7b29ce647b646baa6414790d4d0/fastertransformer/cuda/decoding_kernel_check.cpp#L31

**Comment by gyin-ai** [comment]
@byshiue I have another question. Can we use gpt converted model with beam search in `decoding_sample.py`?

**Comment by byshiue** [comment]
> Thanks. Can we add it as the sampling decoding output firstly? I saw gpt example used sampling. TopK=1 would be fine.
> 
> I noticed we had something mentioned here. If we can expose this variable, it would work as expected.
> 
> https://github.com/NVIDIA/FasterTransformer/blob/93b7751bcbb4b7b29ce647b646baa6414790d4d0/fastertransformer/cuda/decoding_kernel_check.cpp#L31
For feature request, we may update at next releasing.

> @byshiue I have another question about gpt. Can we use gpt converted model with beam search?
Currently not.
